File: examples/zig/oracle-price/OraclePriceFeed_test.zig
const std = @import("std");

const root = @import("../examples_test.zig");
const runar = @import("runar");
const OraclePriceFeed = @import("OraclePriceFeed.runar.zig").OraclePriceFeed;

fn contractPath(comptime basename: []const u8) []const u8 {
    return "oracle-price/" ++ basename;
}

fn runCompileChecks(comptime basename: []const u8) !void {
    try root.runar.compileCheckSource(std.testing.allocator, @embedFile(basename), basename);
    try root.runar.compileCheckFile(std.testing.allocator, contractPath(basename));
}

test "compile-check OraclePriceFeed.runar.zig" {
    try runCompileChecks("OraclePriceFeed.runar.zig");
}

test "OraclePriceFeed init stores oracle and receiver" {
    const contract = OraclePriceFeed.init(&runar.testing.rabin_test_key_n, runar.ALICE.pubKey);
    try std.testing.expectEqualSlices(u8, &runar.testing.rabin_test_key_n, contract.oraclePubKey);
    try std.testing.expectEqualSlices(u8, runar.ALICE.pubKey, contract.receiver);
}

test "OraclePriceFeed settle accepts a real oracle proof above the threshold" {
    const contract = OraclePriceFeed.init(&runar.testing.rabin_test_key_n, runar.ALICE.pubKey);
    const price: i64 = 60_000;
    const proof = runar.testing.oraclePriceProof(price).?;

    contract.settle(price, proof.sig, proof.padding, runar.signTestMessage(runar.ALICE));
}

test "OraclePriceFeed rejects invalid oracle proofs, thresholds, and receiver signatures" {
    try root.expectAssertFailure("oracle-price-wrong-rabin-proof");
    try root.expectAssertFailure("oracle-price-below-threshold");
    try root.expectAssertFailure("oracle-price-wrong-receiver-sig");
}

File: examples/ts/oracle-price/OraclePriceFeed.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract, ALICE, signTestMessage, rabinSign, RABIN_TEST_KEY } from 'runar-testing';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'OraclePriceFeed.runar.ts'), 'utf8');

const ORACLE_PK = RABIN_TEST_KEY.n;
const RECEIVER_SIG = signTestMessage(ALICE.privKey);

/** Encode a positive price as 8-byte unsigned LE (matches num2bin(price, 8n)) */
function priceToMsg(price: bigint): Uint8Array {
  const buf = new Uint8Array(8);
  let p = price;
  for (let i = 0; i < 8 && p > 0n; i++) {
    buf[i] = Number(p & 0xffn);
    p >>= 8n;
  }
  return buf;
}

/** Sign a price with the Rabin test key, returning sig (bigint) and padding (hex string) */
function signPrice(price: bigint): { rabinSig: bigint; padding: string } {
  const { sig, padding } = rabinSign(priceToMsg(price), RABIN_TEST_KEY);
  // Convert padding bigint to hex ByteString (unsigned LE)
  if (padding === 0n) return { rabinSig: sig, padding: '00' };
  const padBytes: number[] = [];
  let p = padding;
  while (p > 0n) { padBytes.push(Number(p & 0xffn)); p >>= 8n; }
  return { rabinSig: sig, padding: padBytes.map(b => b.toString(16).padStart(2, '0')).join('') };
}

describe('OraclePriceFeed', () => {
  function makeOracle() {
    return TestContract.fromSource(source, {
      oraclePubKey: ORACLE_PK,
      receiver: ALICE.pubKey,
    });
  }

  it('settles when price exceeds threshold', () => {
    const oracle = makeOracle();
    const { rabinSig, padding } = signPrice(60000n);
    const result = oracle.call('settle', {
      price: 60000n,
      rabinSig,
      padding,
      sig: RECEIVER_SIG,
    });
    expect(result.success).toBe(true);
  });

  it('rejects settlement when price is below threshold', () => {
    const oracle = makeOracle();
    const { rabinSig, padding } = signPrice(30000n);
    const result = oracle.call('settle', {
      price: 30000n,
      rabinSig,
      padding,
      sig: RECEIVER_SIG,
    });
    expect(result.success).toBe(false);
  });

  it('rejects settlement at exactly the threshold', () => {
    const oracle = makeOracle();
    const { rabinSig, padding } = signPrice(50000n);
    const result = oracle.call('settle', {
      price: 50000n,
      rabinSig,
      padding,
      sig: RECEIVER_SIG,
    });
    // price > 50000n is strict, so 50000n fails
    expect(result.success).toBe(false);
  });
});
